### ISSUE TITLE
전처리 4가지 [의미있는 컬럼, 의미없는 컬럼, PCA, 정규화]

### ISSUE THREAD

**Issue by sangmandu** [sangmandu/Optimus#6]
## 기본적인 분석
- 현재 베이스라인 코드는 1.97의 성능을 냅니다. (seed에 따라 차이 있음)
- 그 중 Y1\~Y7까지가 Y8\~Y14에 비해 높은 NRMSE 점수(점수가 높으면 성능이 안좋은 것임)를 얻습니다.
- 따라서, 이 Y1\~Y7값만을 잘 맞추더라도 좋은 성능을 기대할 수 있을 것입니다.
  - (아래 그림은 lg_nrmse 총합 점수와 Y1~Y14에 대한 각각의 점수를 나타낸 것)
![image](https://user-images.githubusercontent.com/45033215/183115143-4b59bf18-4d18-4987-9959-01949eb68fb4.png)

### 1. 의미없는 컬럼 제거 `완료` ⭕
- 통과여부에 관한 컬럼 `['X_04', 'X_23', 'X_47', 'X_48']` 를 제거하였습니다.

### 2. 의미있는 컬럼 학습 `완료` ⭕
- [Issue 3](https://github.com/sangmandu/Optimus/issues/3)에 종원님이 작성해주신 `y_feature 학습에 가장 큰 영향을 주는 개별 x_feature 탐색하기`의 정보를 이용하였습니다. Y컬럼을 학습하기 위한 X컬럼의 개수를 1개부터 20개까지 학습해보고 성능을 비교해보았습니다.

### 3. PCA `완료` ⭕
![image](https://user-images.githubusercontent.com/45033215/183538749-02c8b634-e8f4-4749-bd07-8ff4856d09cf.png)
- 56개의 x 컬럼을 pca 기법으로 축소해보았습니다.

### 4. Normalize  `완료` ⭕
기존 코드의 LinearRegression 함수에서 normalize 인자를 True로 설정하면 정규화를 할 수 있습니다.

### 1. 의미없는 컬럼 제거
- 당연히 아무런 의미가 없는 컬럼이므로 성능이 증가하지도 감소하지도 않았습니다.

### 2. 의미있는 컬럼 학습
![image](https://user-images.githubusercontent.com/45033215/183442930-bdf847e5-cc84-47c3-8af6-2a44551a78d4.png)
- 전체 56개의 X컬럼 중에서, 최대 20개의 X컬럼을 사용하더라도 비슷한 성능을 내는 것은 확인되었습니다.
- 다만, 그다지 영향력이 없는 X컬럼을 제거한다고 해서 성능이 올라가지는 않았습니다.
![image](https://user-images.githubusercontent.com/45033215/183443329-30905506-105b-421b-b22f-b606edaf4bd4.png)
- 아래에 코드를 남겨놓았습니다

### 3. PCA
![image](https://user-images.githubusercontent.com/45033215/183538947-1cbf609a-84d7-4fca-844d-016f210f1cb0.png)
대체적으로 2점에 가까운 성능을 내어, 기존 베이스라인보다는 좋지 않습니다.
- 아래에 코드를 남겨놓았습니다

### 4. Normalize
약간의 성능 증가를 보입니다.
베이스라인 : 1.977
정규화 : 1.942

![image](https://user-images.githubusercontent.com/45033215/183447773-d92e35a3-3696-418f-bdc8-676a83e98f91.png)

### 2. 의미있는 컬럼 학습
```py
import pandas as pd
import random
import os
import numpy as np

from tqdm import tqdm

from sklearn.linear_model import LinearRegression
from sklearn.multioutput import MultiOutputRegressor
from sklearn import metrics

from sklearn.model_selection import train_test_split

def seed_everything(seed):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)

def lg_nrmse(gt, preds):
    # 각 Y Feature별 NRMSE 총합
    # Y_01 ~ Y_08 까지 20% 가중치 부여
    all_nrmse = []
    for idx in range(14): # ignore 'ID'
        rmse = metrics.mean_squared_error(gt[:,idx], preds[:,idx], squared=False)
        nrmse = rmse/np.mean(np.abs(gt[:,idx]))
        all_nrmse.append(nrmse)
    score = 1.2 * np.sum(all_nrmse[:8]) + 1.0 * np.sum(all_nrmse[8:15])
    return score, all_nrmse

def single_nrmse(gt, preds):
    rmse = metrics.mean_squared_error(gt, preds, squared=False)
    nrmse = rmse/np.mean(np.abs(gt))

    return nrmse

RI = [
    [18, 22, 20, 52, 51, 15, 17, 49, 16, 6, 19, 43, 13, 21, 56, 24, 42, 3, 40, 46],
    [22, 20, 18, 14, 15, 17, 6, 16, 43, 25, 40, 21, 42, 49, 19, 27, 33, 24, 56, 51],
    [18, 22, 43, 19, 49, 20, 17, 13, 21, 16, 6, 24, 28, 5, 51, 45, 56, 52, 42, 41],
    [30, 32, 19, 21, 20, 17, 54, 16, 22, 12, 3, 44, 50, 18, 24, 1, 52, 13, 14, 6],
    [17, 18, 13, 15, 12, 56, 16, 9, 54, 32, 30, 50, 40, 49, 19, 33, 55, 29, 51, 21],
    [19, 30, 17, 18, 22, 16, 6, 56, 43, 5, 49, 50, 9, 51, 26, 3, 27, 44, 13, 55],
    [17, 19, 21, 22, 14, 18, 15, 16, 41, 32, 54, 12, 50, 29, 25, 3, 31, 55, 30, 24],
    [20, 22, 3, 30, 32, 40, 42, 18, 14, 15, 44, 27, 1, 51, 55, 17, 53, 5, 52, 12],
    [20, 22, 30, 3, 42, 18, 32, 40, 15, 51, 13, 27, 1, 55, 53, 52, 5, 12, 26, 56],
    [18, 3, 16, 15, 42, 22, 21, 17, 20, 49, 44, 30, 9, 27, 51, 29, 56, 40, 12, 32],
    [17, 18, 30, 3, 15, 16, 56, 32, 20, 51, 31, 52, 50, 21, 42, 55, 10 ,44, 33, 1],
    [20, 22, 30, 3, 42, 18, 40, 32, 15, 51, 27, 13, 1, 55, 53, 52, 5, 12, 26, 56],
    [20, 22, 30, 3, 42, 18, 40, 32, 15, 51, 27, 1, 13, 55, 53, 52, 5, 12, 26, 10],
    [20, 22, 30, 3, 42, 18, 32, 40, 15, 51, 27, 13, 1, 53, 55, 52, 5, 12, 26, 56],
    ]

RI = [[str(item) if item >= 10 else '0'+str(item) for item in lst] for lst in RI]

best_score = {col: (1, 0) for col in train_y.columns}


for idx, col in enumerate(train_y.columns):
    base_x = pd.DataFrame()
    pbar = tqdm(enumerate(RI[idx], 1))
    for jdx, row in pbar:
        base_x = pd.concat([base_x, train_x[f'X_{row}']], axis=1)

        LR = LinearRegression().fit(base_x, train_y[f'{col}'])
        score = single_nrmse(valid_y[f'{col}'].values, LR.predict(valid_x[list(base_x.columns)]))
            
        if score < best_score[col][0]:
            best_score[col] = (score, row)

        pbar.set_description(f'{jdx} of {col} processing : {score} | columns : {list(base_x.columns)}')

    print(f'{col} : {best_score[col]}')

best_score
```

### 3. PCA
```py
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import plotly.express as px

x = StandardScaler().fit_transform(train_x)

pca = PCA()
pca.fit(x)

exp_var_cumul = np.cumsum(pca.explained_variance_ratio_)

px.area(
    x=range(1, exp_var_cumul.shape[0] + 1),
    y=exp_var_cumul,
    labels={"x": "# Components", "y": "Explained Variance"}
)

norm_train_x = train_x.copy() #StandardScaler().fit_transform(train_x)
norm_valid_x = valid_x.copy() #StandardScaler().fit_transform(valid_x)

for n in range(1, 57):
    pca = PCA(n_components=n)
    
    comp = pca.fit_transform(norm_train_x)
    pca_train_x = pd.DataFrame(data=comp, columns = [f'Z_{(str(idx))}' for idx in range(1, n+1)])
    
    LR = MultiOutputRegressor(LinearRegression()).fit(pca_train_x, train_y)

    comp = pca.fit_transform(norm_valid_x)
    pca_valid_x = pd.DataFrame(data=comp, columns = [f'Z_{(str(idx))}' for idx in range(1, n+1)])

    valid_preds = LR.predict(pca_valid_x)
    print(n, ' : ', lg_nrmse(valid_y.values, valid_preds))

```
